model-sync-lib/src/main/kotlin/org/modelix/model/sync/ModelImporter.kt

```diff
@@ -0,0 +1,99 @@
+
+    private lateinit var originalIdToRef: MutableMap<String, INodeReference>
+
+    fun import(jsonFile: File) {
+        require(jsonFile.exists())
+        require(jsonFile.extension == "json")
+
+        val data = ModelData.fromJson(jsonFile.readText())
+        import(data)
+    }
+    
+    fun import(data: ModelData) {
+        originalIdToRef = mutableMapOf()
+        branch.runWrite {
+            val rootNode = branch.getRootNode()
+            if (rootNode.allChildren.none()) {
+                data.load(branch)
+                return@runWrite
+            }
+            buildRefIndex(rootNode)
+            syncNode(rootNode, data.root)
+        }
+    }
+
+    private fun buildRefIndex(node: INode) {
+        node.originalId()?.let { originalIdToRef[it] = node.reference }
+        node.allChildren.forEach { buildRefIndex(it) }
+    }
+
+    private fun syncNode(node: INode, nodeData: NodeData) {
+        syncProperties(node, nodeData)
+        syncReferences(node, nodeData)
+        syncChildNodes(node, nodeData)
+    }
+
+    private fun syncProperties(node: INode, nodeData: NodeData) {
+        nodeData.properties.forEach { node.setPropertyValue(it.key, it.value) }
+    }
+
+    private fun syncReferences(node: INode, nodeData: NodeData) {
+        nodeData.references.forEach { node.setReferenceTarget(it.key, originalIdToRef[it.value]) }
+    }
+
+    private fun syncChildNodes(node: INode, nodeData: NodeData) {
+        val toBeRemoved = mutableSetOf<INode>()
+        val existingIds = mutableSetOf<String>()
+
+        for (child in node.allChildren) {
+            child.originalId()?.let { existingIds.add(it) }
+            val foundChildData: NodeData? = nodeData.children.find { it.id == child.originalId() }
+            if (foundChildData != null) {
+                syncNode(child, foundChildData)
+            } else {
+                toBeRemoved.add(child)
+            }
+        }
+
+        val toBeAdded = nodeData.children.filter { !existingIds.contains(it.id) }
+
+        toBeAdded.forEach {
+            val index = nodeData.children.indexOf(it)
+            val createdNode = node.addNewChild(it.role, index, it.concept?.let { s -> ConceptReference(s) })
+            createdNode.setPropertyValue(NodeData.idPropertyKey, it.originalId())
+        }
+        toBeRemoved.forEach { node.removeChild(it) }
+            syncChildOrder(node, nodeData)
+    }
+
+    private fun syncChildOrder(node: INode, nodeData: NodeData) {
+        val existingChildren = node.allChildren.toList()
+        val specifiedChildren = nodeData.children
+        require(existingChildren.size == specifiedChildren.size)
+
+        for ((actualNode, specifiedNode) in existingChildren zip specifiedChildren) {
+            val actualId = actualNode.originalId()
+            if (actualId != specifiedNode.originalId() ) {
+                val targetIndex = specifiedChildren.indexOfFirst { actualId == it.originalId() }
```

Comment:
That's the wrong index. moveChild expects the index inside the role, but this is the index across all roles.